feat: JS SDK loader - CDN support

HazAT opened the pull request:
<img width="917" alt="screenshot 2018-06-03 12 48 50" src="https://user-images.githubusercontent.com/363802/40885740-81b413fc-672c-11e8-8900-aeee93e4bb59.png">

This adds support for fastly CDN using our existing js sdk loader template.
It removes the ability to change the URL of the SDK on the fly.
Adds required headers to make fastly CDN work correctly.

Loader is reachable with 
`<sentry_url>/js-sdk-loader/<public_key>.js`
or for the minified version
`<sentry_url>/js-sdk-loader/<public_key>.min.js`

mattrobenolt commented on src/sentry/models/projectkey.py:
```diff
@@ -28,9 +28,11 @@
 
 _uuid4_re = re.compile(r'^[a-f0-9]{32}$')
-
+CDN_URL = 'https://js.sentry-cdn.com/'
```
This needs to be configurable.

mattrobenolt commented on src/sentry/web/frontend/js_sdk_loader.py:
```diff
@@ -0,0 +1,41 @@
+
+    def get(self, request, public_key, minified):
+        """Returns a js file that can be integrated into a website"""
+        key = ProjectKey.objects.get(
```
Need to handle a 404 here.

mattrobenolt commented on src/sentry/web/frontend/js_sdk_loader.py:
```diff
@@ -0,0 +1,41 @@
+        context = {
+            'config': config.get_project_key_config(key),
+            'jsSdkUrl': DEFAULT_SDK_URL,
+            'minified': minified
```
Wouldn't it be simpler to have two separate templates? I imagine potentially we'd want to minify the file during static file compilation in the future.

mattrobenolt commented on src/sentry/web/frontend/js_sdk_loader.py:
```diff
@@ -0,0 +1,41 @@
+
+
+SDK_VERSION = '0.5.2'
+DEFAULT_SDK_URL = 'https://s3.amazonaws.com/getsentry-cdn/@sentry/browser/%s/bundle.min.js' % SDK_VERSION
```
I don't have an answer for this, but this shouldn't be hardcoded. I think we should make this configurable somehow.

HazAT commented:
I moved everything to configure in `sever.py` now.
Before this PR it was possible, for the user, to change the bundle URL, which could be really bad since they in theory could direct to malicious code.
I think it's flexible enough now in the server config but can also move it somewhere else if you think it's better.

HazAT commented:
@mattrobenolt Ready to merge?

mattrobenolt commented on src/sentry/conf/server.py:
```diff
@@ -1343,3 +1344,11 @@ def get_raven_config():
+# If this is an absolute url like e.g.: https://js.sentry-cdn.com/
+# the full url will look like this: https://js.sentry-cdn.com/<public_key>.min.js
+# otherwise django reverse url lookup will be used.
+JS_SDK_LOADER_CDN_URL = ''
```
Don't you want this `JS_SDK_LOADER_CDN_URL = '/js-sdk-loader/'` or something? Otherwise, it'll generate wrong urls.

HazAT replied:
The idea here is to only add a full url that should be used instead of the `reverse` django url see: https://github.com/getsentry/sentry/pull/8640/files/ce88a840cffd2631bd17500b360f101902603752#diff-9d1fb52b257784a8bb2c81b5a0631c76R204

So adding anything instead of a full `http(s)://` url will do nothing, also see your other comment.

mattrobenolt commented on src/sentry/conf/server.py:
```diff
@@ -1343,3 +1344,11 @@ def get_raven_config():
+# otherwise django reverse url lookup will be used.
+JS_SDK_LOADER_CDN_URL = ''
+JS_SDK_LOADER_SDK_VERSION = '0.5.2'
+JS_SDK_LOADER_DEFAULT_SDK_URL = 'https://s3.amazonaws.com/getsentry-cdn/@sentry/browser/%s/bundle.min.js' % JS_SDK_LOADER_SDK_VERSION
```
I assume this is just a placeholder? We're not serving this directly from S3.

HazAT replied:
Ehm, yup this is basically a placeholder, it is a working bundle but in no way final. The loader as it works now can only handle the new JS SDK, so this will be changed.

HazAT commented:
@mattrobenolt 
So I removed the placeholder settings and added a noop template.
We need to configure the stuff then in getsentry/getsentry I suppose.

mattrobenolt commented on src/sentry/models/projectkey.py:
```diff
@@ -191,16 +189,31 @@ def security_endpoint(self):
 
+    @property
+    def js_sdk_loader_cdn_url(self):
+        if bool(_https_re.match(settings.JS_SDK_LOADER_CDN_URL)):
```
I'd really just prefer we don't do this with a regular expression. It's just unnecessary when you can just do a simple `if settings.JS_SDK_LOADER_CDN_URL` instead.

HazAT replied:
@mattrobenolt done ✅
